core/pkg/sync-store/sync_store.go

```diff
@@ -0,0 +1,287 @@
+package store
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"regexp"
+	"sync"
+	"time"
+
+	"github.com/open-feature/flagd/core/pkg/logger"
+	isync "github.com/open-feature/flagd/core/pkg/sync"
+	"github.com/open-feature/flagd/core/pkg/sync/file"
+	"github.com/open-feature/flagd/core/pkg/sync/kubernetes"
+	"go.uber.org/zap"
+)
+
+var (
+	regCrd  *regexp.Regexp
+	regFile *regexp.Regexp
+)
+
+func init() {
+	regCrd = regexp.MustCompile("^core.openfeature.dev/")
+	regFile = regexp.MustCompile("^file:")
+}
+
+type SyncStore struct {
+	ctx          context.Context
+	syncHandlers map[string]*syncHandler
+	logger       *logger.Logger
+	mu           *sync.RWMutex
+	syncBuilder  SyncBuilderInterface
+}
+
+type syncHandler struct {
+	subs       map[interface{}]storedChannels
+	dataSync   chan isync.DataSync
+	cancelFunc context.CancelFunc
+	syncRef    isync.ISync
+	mu         *sync.RWMutex
+}
+
+type storedChannels struct {
+	errChan  chan error
+	dataSync chan isync.DataSync
+}
+
+func NewSyncStore(ctx context.Context, logger *logger.Logger) *SyncStore {
+	ss := SyncStore{
+		ctx:          ctx,
+		syncHandlers: map[string]*syncHandler{},
+		logger:       logger,
+		mu:           &sync.RWMutex{},
+		syncBuilder:  &SyncBuilder{},
+	}
+	go ss.cleanup()
+	return &ss
+}
+
+func (s *SyncStore) FetchAllFlags(ctx context.Context, key interface{}, target string) (isync.DataSync, error) {
+	s.logger.Debug(fmt.Sprintf("fetching all flags for target %s", target))
+	dataSyncChan := make(chan isync.DataSync, 1)
+	errChan := make(chan error, 1)
+	s.mu.RLock()
+	syncHandler, ok := s.syncHandlers[target]
+	s.mu.RUnlock()
+	if !ok {
+		s.logger.Debug(fmt.Sprintf("sync handler does not exist for target %s, registering a new subscription", target))
+		s.RegisterSubscription(ctx, target, key, dataSyncChan, errChan)
+	} else {
+		if syncHandler.syncRef == nil {
+			return isync.DataSync{}, errors.New("sync ref not set")
+		}
+		go func() {
+			s.logger.Debug(fmt.Sprintf("sync handler exists for target %s, triggering a resync", target))
+			if err := syncHandler.syncRef.ReSync(ctx, dataSyncChan); err != nil {
+				errChan <- err
+			}
+		}()
+	}
+
+	select {
+	case data := <-dataSyncChan:
+		return data, nil
+	case err := <-errChan:
+		return isync.DataSync{}, err
+	case <-time.After(5 * time.Second):
+		return isync.DataSync{}, errors.New("fetching all flags timed out after 5 seconds")
+	}
+}
+
+func (s *SyncStore) RegisterSubscription(
+	ctx context.Context,
+	target string,
+	key interface{},
```

Comment:
Do we need this `key` 🤔  I still couldn't figure out why we need it. May be I am missing the usage

Reply:
its a unique identifier for the subscription channel that we can use without allocating additional memory, its just in the subscription map -> it also gives us a clean trace ID if we choose to go that way